// reflex/lib/index.js
(function (global, factory) {
  if (typeof define === 'function' && define.amd) {
    define(['exports', 'react'], factory);
  } else if (typeof exports !== 'undefined') {
    factory(exports, require('react'));
  } else {
    var mod = {
      exports: {}
    };
    factory(mod.exports, global.React);
    global.index = mod.exports;
  }
})(this, function (exports, _react) {
  'use strict';

  var _get = function get(_x5, _x6, _x7) { var _again = true; _function: while (_again) { var object = _x5, property = _x6, receiver = _x7; desc = parent = getter = undefined; _again = false; if (object === null) object = Function.prototype; var desc = Object.getOwnPropertyDescriptor(object, property); if (desc === undefined) { var parent = Object.getPrototypeOf(object); if (parent === null) { return undefined; } else { _x5 = parent; _x6 = property; _x7 = receiver; _again = true; continue _function; } } else if ('value' in desc) { return desc.value; } else { var getter = desc.get; if (getter === undefined) { return undefined; } return getter.call(receiver); } } };

  var _createClass = (function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ('value' in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; })();

  function _toConsumableArray(arr) { if (Array.isArray(arr)) { for (var i = 0, arr2 = Array(arr.length); i < arr.length; i++) arr2[i] = arr[i]; return arr2; } else { return Array.from(arr); } }

  function _defaults(obj, defaults) { var keys = Object.getOwnPropertyNames(defaults); for (var i = 0; i < keys.length; i++) { var key = keys[i]; var value = Object.getOwnPropertyDescriptor(defaults, key); if (value && value.configurable && obj[key] === undefined) { Object.defineProperty(obj, key, value); } } return obj; }

  function _inherits(subClass, superClass) { if (typeof superClass !== 'function' && superClass !== null) { throw new TypeError('Super expression must either be null or a function, not ' + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) _defaults(subClass, superClass); }

  function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

  function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError('Cannot call a class as a function'); } }

  if (typeof Symbol === 'undefined') {
    var Symbol = function Symbol(name) {
      return '@@' + name + '#' + Math.random().toString(36).substr(2);
    };
    Symbol['for'] = function (name) {
      return '@@' + name;
    };
  }

  if (typeof WeakMap === 'undefined') {
    var WeakMap;

    (function () {
      var weakMapID = 0;

      WeakMap = (function () {
        var _class = function WeakMap() {
          _classCallCheck(this, _class);

          this.id = 'weak#' + (++weakMapID).toString(36);
        };

        _createClass(_class, [{
          key: 'has',
          value: function has(key) {
            return this.hasOwnProperty.call(key, this.id);
          }
        }, {
          key: 'set',
          value: function set(key, value) {
            key[this.id] = value;
          }
        }, {
          key: 'get',
          value: function get(key) {
            return key[this.id];
          }
        }, {
          key: 'delete',
          value: function _delete(key) {
            delete key[this.id];
          }
        }, {
          key: 'clear',
          value: function clear() {
            this.constructor();
            console.warn('WeakMap polyfill does not release references');
          }
        }]);

        return _class;
      })();
    })();
  }

  // node function constructs a virtual dom nodes. It takes node `tagName`,
  // optional `model` properties and optional `children` and produces instance
  // of the appropriate virtual `Node`.
  var node = function node(tagName, model) {
    var children = arguments[2] === undefined ? NodeList.empty : arguments[2];

    var properties = {};

    if (model) {
      for (var _name in model) {
        if (model.hasOwnProperty(_name)) {
          var value = model[_name];
          if (_name === 'style' && value && value.toJSON) {
            properties[_name] = value.toJSON();
          } else {
            properties[_name] = value;
          }
        }
      }
    }

    properties.children = children;

    return _react.createElement(tagName, properties);
  };

  exports.node = node;
  // Create a function for react supported nodes.
  var html = Object.create(null);
  exports.html = html;
  Object.keys(_react.DOM).forEach(function (tagName) {
    html[tagName] = function (model, children) {
      return node(tagName, model, children);
    };
  });

  // Fragment insntances represents a namespacing facility for
  // entities in virtual dom tree, primarily to group several keyed
  // nodes without wrapping them in an element. This allows different
  // groups of nodes to be joined under the same parent. Namespacing
  // such groups with a `properties.key` is crucial for optimising
  // performance of virtual dom diffing algorithm. Optionally
  // `properties.type` can be passed that will be a tagName of node
  // if fragment is rendered as a root of dom tree, or root of the
  // subtree that is computed lazily by a thunk. `children` argument
  // is an array of nodes that this fragment represents.

  var Fragment = function Fragment(properties, children) {
    _classCallCheck(this, Fragment);

    this.key = properties.key;
    this.type = properties.type || 'x:fragment';
    this.properties = properties;
    this.children = children;

    this._reactFragment = _defineProperty({}, this.key, this.children);
  };

  // Function for constructing fragments of virtual dom tree.
  // Takes `properties.key` for identifying fragment, optional
  // `properties.type` which is used as wrapper node for given
  // `children` if fragment is returned from thunk or is rendered
  // directly.
  var fragment = function fragment(properties, children) {
    return new Fragment(properties, children);
  };

  exports.fragment = fragment;

  var Forwarder = (function () {
    function Forwarder(addressBook, to) {
      _classCallCheck(this, Forwarder);

      this.addressBook = addressBook;
      this.to = to;
    }

    _createClass(Forwarder, [{
      key: 'receive',
      value: function receive(action) {
        this.addressBook[this.to].receive(action);
      }
    }]);

    return Forwarder;
  })();

  // Thunk instances are entities in a virtual dom tree that represent
  // lazily compute sub-trees with built-in caching that avoid re-computing
  // sub-tree when same data is being rendered.

  var Thunk = (function (_React$Component) {
    function Thunk(props, context) {
      _classCallCheck(this, Thunk);

      _get(Object.getPrototypeOf(Thunk.prototype), 'constructor', this).call(this, props, context);
    }

    _inherits(Thunk, _React$Component);

    _createClass(Thunk, [{
      key: 'componentWillMount',
      value: function componentWillMount() {
        var args = this.props.args;

        var count = args.length;

        var addressBook = new Array(count);
        var params = new Array(count);

        var index = 0;
        while (index < count) {
          var arg = args[index];
          if (arg instanceof Address) {
            addressBook[index] = arg;
            params[index] = new Address(new Forwarder(addressBook, index));
          } else {
            params[index] = arg;
          }
          index = index + 1;
        }

        this.setState({ args: params, addressBook: addressBook });
      }
    }, {
      key: 'componentWillReceiveProps',
      value: function componentWillReceiveProps(_ref) {
        var after = _ref.args;

        if (profile) {
          console.time(this.props.Key + '.compute');
        }

        var _state = this.state;
        var args = _state.args;
        var addressBook = _state.addressBook;

        var count = after.length;
        var index = 0;
        var isUpdated = false;

        if (args.length !== count) {
          isUpdated = true;
          args.length = count;
          addressBook.length = count;
        }

        while (index < count) {
          var next = after[index];
          var arg = args[index];

          if (next === arg) {} else if (next instanceof Address && arg instanceof Address) {
            // Update adrress book with a new address.
            addressBook[index] = next;
          } else {
            isUpdated = true;

            if (next instanceof Address) {
              addressBook[index] = next;
              args[index] = new Address(this, [redirect(index)]);
            } else {
              args[index] = next;
            }
          }
          index = index + 1;
        }

        if (isUpdated) {
          this.setState({ args: args, addressBook: addressBook });
        }

        if (profile) {
          console.timeEnd(this.props.Key + '.compute');
        }
      }
    }, {
      key: 'shouldComponentUpdate',
      value: function shouldComponentUpdate(_, state) {
        return state !== this.state;
      }
    }, {
      key: 'render',
      value: function render() {
        if (profile) {
          console.time(this.props.Key + '.render');
        }

        // Store currently operating view to enable cacheing by
        // the view.
        Thunk.context = this.view;

        var args = this.state.args;

        var result = this.view.apply(this, _toConsumableArray(args));

        Thunk.context = null;

        if (profile) {
          console.timeEnd(this.props.Key + '.render');
        }
        return result;
      }
    }], [{
      key: 'for',
      value: function _for(view, key) {
        var ReactComponent = (function (_Thunk) {
          function ReactComponent(props, context) {
            _classCallCheck(this, ReactComponent);

            _get(Object.getPrototypeOf(ReactComponent.prototype), 'constructor', this).call(this, props, context);
            this.view = view;
          }

          _inherits(ReactComponent, _Thunk);

          return ReactComponent;
        })(Thunk);

        ReactComponent.displayName = key.split('@')[0];
        return ReactComponent;
      }
    }]);

    return Thunk;
  })(_react.Component);

  var contextualCache = new WeakMap();
  var contextlessCache = new WeakMap();

  var cache = function cache(f) {
    for (var _len = arguments.length, args = Array(_len > 1 ? _len - 1 : 0), _key = 1; _key < _len; _key++) {
      args[_key - 1] = arguments[_key];
    }

    var cache = null;
    var changed = false;
    var context = Thunk.context;

    if (context) {
      cache = contextualCache.get(f);
      if (!cache) {
        contextualCache.set(f, cache = new WeakMap());
      }
    } else {
      cache = contextlessCache;
    }

    var memory = cache.get(f);
    if (memory) {
      var input = memory.input;

      var count = input.length;
      var index = 0;
      var _changed = args.length !== count;
      while (!_changed && index < count && !_changed) {
        var past = input[index];
        var current = args[index];

        _changed = current !== past;
        index = index + 1;
      }

      if (_changed) {
        memory.output = f.apply(undefined, args);
        memory.input = args;
      }
    } else {
      memory = { output: f.apply(undefined, args), input: args };
      cache.set(f, memory);
    }

    return memory.output;
  };

  exports.cache = cache;
  // render function provides shortcut for rendering a model with
  // default view function (although custom view function can be
  // passed as second optional argument), but unlike calling view
  // directly result is a thunk, there for it's defers actual computation
  // and makes use of caching to avoid computation when possible.
  var render = function render(key, view) {
    for (var _len2 = arguments.length, args = Array(_len2 > 2 ? _len2 - 2 : 0), _key2 = 2; _key2 < _len2; _key2++) {
      args[_key2 - 2] = arguments[_key2];
    }

    var component = view.reactComponent || (view.reactComponent = Thunk['for'](view, key));
    return _react.createElement(component, { key: key, Key: key, args: args });
  };

  exports.render = render;
  var GUID = 0;

  var ParameterizedReceiver = function ParameterizedReceiver(address, read, prefix) {
    return function (event) {
      return address.receive(read.apply(undefined, _toConsumableArray(prefix).concat([event])));
    };
  };

  var Receiver = function Receiver(address, read) {
    return function (event) {
      return address.receive(read(event));
    };
  };

  var Address = (function () {
    function Address(mailbox, forwarders) {
      _classCallCheck(this, Address);

      this.id = ++GUID;
      this.mailbox = mailbox;
      this.forwarders = forwarders;
    }

    _createClass(Address, [{
      key: 'forward',
      value: function forward(anotate) {
        var cache = anotate[Address.cache] || (anotate[Address.cache] = {});

        if (!cache[this.id]) {
          var forwarders = this.forwarders ? [anotate].concat(this.forwarders) : [anotate];
          cache[this.id] = new Address(this.mailbox, forwarders);
        }

        return cache[this.id];
      }
    }, {
      key: 'receive',
      value: function receive(action) {
        if (this.isBlocked && action !== null) {
          (this.queue || (this.queue = [])).push(action);
        } else {
          var forwarders = this.forwarders;

          this.isBlocked = true;
          var ticket = -1;

          // Define a `delivered` flag that is updated at the begining of action
          // delivery and at the end of it. This is so that in finally we can
          // figure out if exception was throw or not.
          var delivered = false;

          while (this.isBlocked) {
            try {
              delivered = false;

              if (action !== null) {
                if (forwarders) {
                  var count = forwarders.length;
                  var index = 0;
                  while (index < count) {
                    action = forwarders[index](action);
                    index = index + 1;
                  }
                }

                this.mailbox.receive(action);

                delivered = true;
              }
            } finally {
              ticket = ticket + 1;
              this.isBlocked = this.queue && this.queue.length > ticket;
              action = this.isBlocked && this.queue[ticket];

              // If failed to deliver (exception was thrown) and address is still
              // blocked that means there are still pending actions to process. In
              // that case we unblock an address remove delivered actions from a
              // queue and re-enter receive loop. This way receive drains action
              // action loop regardless of exceptions that may occur.
              if (!delivered && this.isBlocked) {
                this.isBlocked = false;
                this.queue && this.queue.splice(0, ticket + 1);
                this.receive(action);
              }
            }
          }

          this.queue && this.queue.splice(0);
        }
      }
    }, {
      key: 'send',
      value: function send(action) {
        var _this = this;

        return function (_) {
          return _this.receive(action);
        };
      }
    }, {
      key: 'pass',
      value: function pass(read) {
        if (typeof read !== 'function') {
          throw TypeError('Non function was passed to address.pass');
        }

        for (var _len3 = arguments.length, prefix = Array(_len3 > 1 ? _len3 - 1 : 0), _key3 = 1; _key3 < _len3; _key3++) {
          prefix[_key3 - 1] = arguments[_key3];
        }

        return prefix.length ? ParameterizedReceiver(this, read, prefix) : read['reflex/address-' + this.id] || (read['reflex/address-' + this.id] = Receiver(this, read));
      }
    }]);

    return Address;
  })();

  exports.Address = Address;

  Address.cache = Symbol['for']('reflex/address-book');

  // Program is a root entity of the virtual dom tree that is responsible
  // for computing a virtual dom tree for the `state` via given `view` function
  // and reacting to dispatched actions via given `update` by updating a state
  // and restarting a the same cycle again.

  var Application = (function () {
    function Application(_ref2) {
      var address = _ref2.address;
      var target = _ref2.target;
      var state = _ref2.state;
      var update = _ref2.update;
      var view = _ref2.view;

      _classCallCheck(this, Application);

      this.target = target;
      this.view = view;
      this.update = update;
      this.state = state;
      this.address = address;
      this.render = this.render.bind(this);
    }

    _createClass(Application, [{
      key: 'receive',
      value: function receive(action) {
        this.action = action;
        var state = this.update(this.state, this.action);
        if (state !== this.state) {
          this.state = state;
          this.schedule();
        }
      }
    }, {
      key: 'schedule',
      value: function schedule() {
        if (!this.isScheduled) {
          this.isScheduled = true;
          this.version = requestAnimationFrame(this.render);
        }
      }
    }, {
      key: 'render',
      value: function render() {
        if (profile) {
          console.time('React.render');
        }

        var start = performance.now();

        // It is important to mark `isScheduled` as `false` before doing actual
        // rendering since state changes in effect of reflecting current state
        // won't be handled by this render cycle. For example rendering a state
        // with updated focus will cause `blur` & `focus` events to be dispatched
        // that happen synchronously, and there for another render cycle may be
        // scheduled for which `isScheduled` must be `false`. Attempt to render
        // this state may also cause a runtime exception but even then we would
        // rather attempt to render updated states that end up being blocked
        // forever.
        this.isScheduled = false;
        if (profile) {
          console.time('Application.render');
        }

        this.tree = this.view(this.state, this.address);

        if (profile) {
          console.timeEnd('Application.render');
        }

        if (profile) {
          console.time('React.reconcile');
        }

        _react.render(this.tree, this.target);

        if (profile) {
          console.timeEnd('React.reconcile');
        }

        var end = performance.now();
        var time = end - start;

        if (time > 16) {
          console.warn('Render took ' + time + 'ms & will cause frame drop');
        }

        if (profile) {
          console.timeEnd('React.render');
        }
        return this;
      }
    }]);

    return Application;
  })();

  exports.Application = Application;

  // Function that takes `target` element to continiusly render given `model`
  // into. model's default `update` function is used to update module in response
  // to dispatched actions and model's default `view` function is used to compute
  // dom tree represantation of the model. Optionally custom `update` and `view`
  // functions could be passed to customize render loop.
  var main = function main(target, model) {
    var update = arguments[2] === undefined ? model.constructor.update : arguments[2];
    var view = arguments[3] === undefined ? model.constructor.view : arguments[3];
    return (function () {
      var application = new Application({
        state: model,
        target: target, update: update, view: view
      });
      application.address = new Address(application);

      application.schedule();
      return application;
    })();
  };

  exports.main = main;
  var profile = null;
  var time = function time() {
    var name = arguments[0] === undefined ? '' : arguments[0];
    return profile = name + ' ';
  };
  exports.time = time;
  var timeEnd = function timeEnd() {
    return profile = null;
  };
  exports.timeEnd = timeEnd;
});

// do nothing
